### PR TITLE
Apollo4 adc driver

### PR DESCRIPTION
**Pull request by RichardSWheatley**
Update Ambiq ADC Driver to add apollo4 plus

Signed-off-by: Richard Wheatley <richard.wheatley@ambiq.com>

### CODE REVIEW

**Comment by RichardSWheatley** [comment]
Converted to draft until https://github.com/zephyrproject-rtos/zephyr/pull/73763 is finished.
Then I can rebase and add my changes here.

**Comment by RichardSWheatley** [comment]
@decsny please see my comments on https://github.com/zephyrproject-rtos/zephyr/pull/73763 addressing your concerns there.

**Comment by RichardSWheatley** [comment]
Updated due to clang-format request from Zephyr-bot.

**Comment by RichardSWheatley** [comment]
Added ASYNC for apollo3/4 and PM in for apollo4 plus

File: drivers/adc/adc_ambiq.c [hunk 1]
```diff
@@ -343,47 +384,56 @@ static int adc_ambiq_pm_action(const struct device *dev, enum pm_device_action a
...
+#endif
 
-#define ADC_AMBIQ_INIT(n)                                                                  \
-	PINCTRL_DT_INST_DEFINE(n);                                                             \
-	ADC_AMBIQ_DRIVER_API(n);                                                               \
-	static int pwr_on_ambiq_adc_##n(void)                                                  \
-	{                                                                                      \
+#define ADC_AMBIQ_INIT(n)                                                                          \
+	PINCTRL_DT_INST_DEFINE(n);                                                                 \
+	ADC_AMBIQ_DRIVER_API(n);                                                                   \
```
Comment (decsny) [inline drivers/adc/adc_ambiq.c:406]: please do format updates in a separate commit

Comment (RichardSWheatley) [reply drivers/adc/adc_ambiq.c:406]: I will fix tomorrow. Thank you!

File: drivers/adc/adc_ambiq.c
Comment (decsny) [inline drivers/adc/adc_ambiq.c:1]: the commit message on eda385da95c5b3967b1e685a5e6a51820071f505 should be improved, don't link to a github "conversation", instead, clearly list all the things the commit is doing and why

File: drivers/adc/adc_ambiq.c [hunk 2]
```diff
@@ -157,9 +162,15 @@ static int adc_ambiq_start_read(const struct device *dev, const struct adc_seque
 		LOG_ERR("No channel selected");
 		return -EINVAL;
 	}
+
+	error = adc_ambiq_check_buffer_size(sequence, active_channels);
```
Comment (decsny) [inline drivers/adc/adc_ambiq.c:166]: for example, why has this been moved

**Review by decsny — COMMENTED** [review]
to me the last 4 commits seem fine, just the first commit to the driver please make more clear in the message what it does

**Comment by RichardSWheatley** [comment]
Can anyone tell me why the twister talent failed with the same code it passed with for the last week?

**Comment by aaronyegx** [comment]
> Can anyone tell me why the twister talent failed with the same code it passed with for the last week?

See the errors:

/__w/zephyr/zephyr/drivers/serial/uart_pl011_ambiq.h:74:12: error: 'uart_ambiq_pm_action' defined but not used [-Werror=unused-function]
   74 | static int uart_ambiq_pm_action(const struct device *dev, enum pm_device_action action)
      |            ^~~~~~~~~~~~~~~~~~~~
cc1: all warnings being treated as errors

It seems the uart_ambiq_pm_action() is only used by Apollo3, which includes in https://github.com/zephyrproject-rtos/zephyr/pull/77110

**Comment by RichardSWheatley** [comment]
> > Can anyone tell me why the twister talent failed with the same code it passed with for the last week?
> 
> 
> 
> See the errors:
> 
> 
> 
> /__w/zephyr/zephyr/drivers/serial/uart_pl011_ambiq.h:74:12: error: 'uart_ambiq_pm_action' defined but not used [-Werror=unused-function]
> 
>    74 | static int uart_ambiq_pm_action(const struct device *dev, enum pm_device_action action)
> 
>       |            ^~~~~~~~~~~~~~~~~~~~
> 
> cc1: all warnings being treated as errors
> 
> 
> 
> It seems the uart_ambiq_pm_action() is only used by Apollo3, which includes in https://github.com/zephyrproject-rtos/zephyr/pull/77110

So, only the apollo3 UART PM was addresssed in your update?

I will have to address that first.

**Comment by RichardSWheatley** [comment]
> > Can anyone tell me why the twister talent failed with the same code it passed with for the last week?
> 
> See the errors:
> 
> /__w/zephyr/zephyr/drivers/serial/uart_pl011_ambiq.h:74:12: error: 'uart_ambiq_pm_action' defined but not used [-Werror=unused-function] 74 | static int uart_ambiq_pm_action(const struct device *dev, enum pm_device_action action) | ^~~~~~~~~~~~~~~~~~~~ cc1: all warnings being treated as errors
> 
> It seems the uart_ambiq_pm_action() is only used by Apollo3, which includes in #77110

@decsny Would I be able to fix this issue in this PR or will I need to create another?

**Comment by RichardSWheatley** [comment]
This is now blocked by: https://github.com/zephyrproject-rtos/zephyr/pull/77477
The fix for the UART error created when apollo3 UART PM update was merged.

**Review by AlessandroLuo — APPROVED** [review]

**Review by decsny — APPROVED** [review]
